### PR TITLE
Multiline top level statement, addresses #893

### PR DESCRIPTION
**Pull request by olafurpg**
Supersedes #900

### CODE REVIEW

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 1]
```diff
@@ -144,4 +158,117 @@ object A {
...
+      4
+    }
+  }
+}
+>>>
+object a {
+
+  def foo = {
+
+    def b = {
```
Comment (olafurpg) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: @Daxten @dwijnand do you think a newline should be forced here? We could make it stop for nested defs, but I'm not sure if that makes more sense

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 2]
```diff
@@ -144,4 +158,117 @@ object A {
...
+  }
+  def bar = 2
+}
+>>>
+object a {
+
+  def foo = {
+    2
+  }
+  def bar = 2
```
Comment (olafurpg) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: IMO, there should be a newline before `bar` because `foo` is multiline, what do you think @Daxten @dwijnand ?

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat (hunk 1 above)
Comment (Daxten) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: I think the question is if the first definition inside (anything) should have a newline like:
```
object A {
  object B {

  }
}
```

I think for anything with lots of pameters it would look better to have a newline before the childs start:
```
class A (val a: Int. 
              val a: Int. 
              val a: Int. 
              val a: Int. 
              val a: Int
            ) {

     class B (val a: Int. 
              val a: Int. 
              val a: Int. 
              val a: Int. 
              val a: Int
            ) 
}
```

so maybe just add it generally?

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat (hunk 2 above)
Comment (Daxten) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: Yes!

Comment (olafurpg) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: OK, we can leave that one for a separate PR. I suspect it will not be too hard to accommodate customizations like this one.
 
I'm gonna quickly merge this PR to unblock #898, which moves almost all source files to a new directory layout.

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat (hunk 1 above)
Comment (olafurpg) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: This new scheme of detecting top level statements is very easy to extend. The trick is to replace this `collect` here https://github.com/scalameta/scalafmt/pull/904/files#diff-4f68f079630ceb866ab2d90e4eacf3e3R182 with a custom `Traverser`. That will make it possible to for example exclude the first statement of `Template`.

We can leave this for a separate PR.

**Comment by olafurpg** [comment]
I'm gonna go ahead and merge this to unblock #898 , I've been planning that PR forever and I was determined to get it in this weekend 😄

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 3]
```diff
@@ -34,29 +37,37 @@ package a {
...
-
+<<< single line
+object a {
+case class C1
+sealed trait T1
+final class FC1
+}
+>>>
+object a {
+  case class C1
```
Comment (olafurpg) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: BTW @Daxten , following @dwijnand request, I changed the behavior to not force blank lines on single-line class/object/trait. I think this makes sense for ADTs,

```scala
sealed trait MyEnum
object MyEnum {
  case class A() extends MyEnum
  case object Foo extends MyEnum
}
```

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 4]
```diff
@@ -144,4 +158,117 @@ object A {
...
+}
+<<< Double annotation
+object a {
+  @annot
+  @bannot
+  class A
+}
+>>>
+object a {
+
```
Comment (olafurpg) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: @Daxten Before this PR, iirc, this newline would not have been inserted, right? If so, I'm happy to follow up with another PR changing it back to the old behavior.

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat (hunk 1 above)
Comment (dwijnand) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: Either don't force or force when the previous declaration is multi-line sounds good to me.

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat (hunk 2 above)
Comment (dwijnand) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: Yes!

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 5]
```diff
@@ -34,29 +37,37 @@ package a {
...
-      case class C1
-
-      sealed trait T1
-
-      final class FC1
-
+      val v3 = {
+        3
+      }
       def a = v1
```
Comment (dwijnand) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: I expect a NL here

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 6]
```diff
@@ -124,7 +135,6 @@ object A {
 
   /** test */
   def foo(): Int = 0
-
```
Comment (dwijnand) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: I expect this newline. A comment/scaladoc above a single-line makes it "multi-line", for me.

File: core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat [hunk 7]
```diff
@@ -144,4 +158,117 @@ object A {
...
+      4
+    }
+  }
+}
+<<< single line following multiline (up for debate)
+object a {
+  def foo = {
+    2
+  }
+  def bar = 2
```
Comment (dwijnand) [inline core/src/test/resources/newlines/alwaysBeforeTopLevelStatements.stat]: for the same reason below, there should be a NL here because foo is multi-line
